index.d.ts

```diff
@@ -23,9 +23,20 @@ import { Readable } from 'stream'
 type Mercurius = typeof mercurius
 
 declare namespace mercurius {
-  export interface PubSub {
-    subscribe<TResult = any>(topics: string | string[]): Promise<Readable & AsyncIterableIterator<TResult>>;
-    publish<TResult = any>(event: { topic: string; payload: TResult }, callback?: () => void): void;
+  export type PubSubPublishArgsByKey = {
+    [key: string]: any;
+  };
+
+  export interface PubSub<
+    TPubSubPublishArgsByKey extends PubSubPublishArgsByKey,
+  > {
+      subscribe<TKey extends Extract<keyof TPubSubPublishArgsByKey, string>>(
+        topics: TKey | TKey[],
+      ): Promise<Readable & AsyncIterableIterator<TPubSubPublishArgsByKey[TKey]>>;
+      publish<TKey extends Extract<keyof TPubSubPublishArgsByKey, string>>(
+        event: { topic: TKey; payload: TPubSubPublishArgsByKey[TKey] },
+        callback?: () => void,
+      ): void;
```

Comment:
The `subscribe` is [a wrapper](https://github.com/mercurius-js/mercurius/blob/a45fd77b7529cd763fba367b2f21eb0c583e5fd6/lib/subscriber.js#L21-L27) for the  `emitter` / `mqemitter` [method `on`](https://github.com/mcollina/mqemitter/blob/6b2e908fa6cd002d7b2227211512f7cc0e6f710d/types/mqemitter.d.ts#L16) and `publish` is [a wrapper](https://github.com/mercurius-js/mercurius/blob/a45fd77b7529cd763fba367b2f21eb0c583e5fd6/lib/subscriber.js#L34) for the `emitter` / `mqemitter` [method `emit`](https://github.com/mcollina/mqemitter/blob/6b2e908fa6cd002d7b2227211512f7cc0e6f710d/types/mqemitter.d.ts#L17): 

```ts
export type Message = Record<string, any> & { topic: string }
```

```ts
on(topic: string, listener: (message: Message, done: () => void) => void, callback?: () => void): this
emit(message: Message, callback?: (error?: Error) => void): void
```

There is some mismatch here, both in the old and in the new types. Eg: the publish callback should be: `callback?: (error?: Error) => void`

I think the `emitter` option of the constructor should also be specified to ensure a match between it and what's sent here: https://github.com/mercurius-js/mercurius/blob/a45fd77b7529cd763fba367b2f21eb0c583e5fd6/index.d.ts#L449

(Ideally the `publish()` would be reimplemented as `public(topic, payload, [callback])` and have some way to resolve to a promise rather than the callback, so that an `await pubsub.publish('topic1', 'value1`)` would be possible, but that's out of scope of this PR)

Reply:
Yeah, the internal types are ambiguous and incorrectly mapped by the library authors but it doesn't affect this pull request.